src/main/java/hudson/remoting/Engine.java

```diff
@@ -303,10 +303,12 @@ public void run() {
 
                 // If no protocol worked.
                 if (channel == null) {
+                    final String rejected = "The server rejected the connection: ";
                     if (triedAtLeastOneProtocol) {
-                        onConnectionRejected("None of the protocols were accepted");
+                        events.status(rejected + "None of the protocols were accepted");
+                        Thread.sleep(10 * 1000);
                     } else {
-                        onConnectionRejected("None of the protocols are enabled");
+                        events.error(new Exception(rejected + "None of the protocols were enabled"));
```

Comment:
The observed behavior is that the master will reject connections from a client when it falsely believes the client is already connected. This state on the master will change after it realizes the client is disconnected and a retry will succeed.

When none of the protocols are enabled an error event is fired. One of the listeners will exit the process as a result. The 'return' is effectively dead code. I found that a bit confusing when trying to debug our Jenkins agent connection reliability.